packages/app/components/Questions/Queue/JoinTagGroupButton.tsx

```diff
@@ -0,0 +1,220 @@
+              studentQuestion.text,
+              newQuestionTypes,
+              studentQuestion.groupable,
+              studentQuestion.isTaskQuestion,
+              studentQuestion.location,
+            )
+          }
+        } else if (taskId) {
+          // basically the same thing, but with tasks
+          if (!studentDemo) {
+            // tasks get put as "Mark "taskId" "taskId2" and have no questionTypes
+            await createQuestion(`Mark "${taskId}"`, [], false, true)
+          } else {
+            const newQuestionText = studentDemo.text + ` "${taskId}"`
+            await updateQuestion(
+              newQuestionText,
+              [],
+              studentDemo.groupable,
+              studentDemo.isTaskQuestion,
+              studentDemo.location,
+            )
+          }
+        } else {
+          console.log(
+            'Error: This component was used incorrectly. No questionType or taskId provided',
+          )
+        }
+      } else {
+        // If leaving the tag group, remove the questionType (for tag) or remove off the task from the question text
+        if (questionType) {
+          if (!studentQuestion) {
+            console.log('Error: Student does not have a question to leave')
+            return
+          }
+          if (
+            studentQuestion.text === '' &&
+            studentQuestion.questionTypes.length === 1
+          ) {
+            await leaveQueue(false)
+          } else {
+            const newQuestionTypes = studentQuestion.questionTypes.filter(
+              (qt) => qt.id !== questionType.id,
+            )
+            await updateQuestion(
+              studentQuestion.text,
+              newQuestionTypes,
+              studentQuestion.groupable,
+              studentQuestion.isTaskQuestion,
+              studentQuestion.location,
+            )
+          }
+        } else if (taskId) {
+          if (!studentDemo) {
+            console.log('Error: Student does not have a task to leave')
+            return
+          }
+          if (studentDemo.text === `Mark "${taskId}"`) {
+            await leaveQueue(true)
+          } else {
+            const newQuestionText = studentDemo.text.replace(` "${taskId}"`, '')
+            await updateQuestion(
+              newQuestionText,
+              [],
+              studentDemo.groupable,
+              studentDemo.isTaskQuestion,
+              studentDemo.location,
+            )
+          }
+        } else {
+          console.log(
+            'Error: This component was used incorrectly. No questionType or taskId provided',
+          )
+        }
+      }
+      setIsJoined(!isJoined)
+    } catch (e) {
+      message.error(e.response?.data?.message || 'An unexpected error occurred')
+    }
+    setIsLoading(false)
+  }
+
+  return (
+    <CustomButton
+      size="small"
+      className={isDone ? '' : 'join-or-leave-tag-group-button'}
+      colorstyle={isDone ? 'Done' : isJoined ? 'Leave' : 'Join'}
```

Comment:
Is there a better way to do this? I feel like nester ternary operators not great

Reply:
I don't think there's a better way of doing it. Also nested ternary operators are used in tons of places in the frontend (usually for whether or not to render certain components).